examples/02_intermediate/wikimedia.ipynb

```diff
@@ -0,0 +1,698 @@
+    "    # Update Perspective widget with each event/tick\n",
+    "    event = update_widget(result, widget=widget)\n",
+    "    # We can also print the event object\n",
+    "    csp.print(\"Event\", event)\n",
+    "    # Add this edge as a graph output\n",
+    "    csp.add_graph_output(\"Wiki events\", event)\n",
+    "    print(\"End of graph building\")\n",
+    "\n",
+    "start = datetime.utcnow()\n",
+    "csp.run(wiki_graph, starttime=start, endtime=start+timedelta(seconds=30), realtime=True)\n",
+    "print(\"Done.\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f707e182-df6e-4802-af78-3f691f0fdf42",
+   "metadata": {},
+   "source": [
+    "## Using CSP to read historical data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4104ae2f-defd-41a3-85ae-30f2c2271b3e",
+   "metadata": {},
+   "source": [
+    "Now, let's explore writing a _historical_ adapter to read the same data stream. The Wikimedia API offers the following query for historial data:\n",
+    "\n",
+    "```\n",
+    "URL = f'https://stream.wikimedia.org/v2/stream/recentchange?since={start_date}'\n",
+    "```\n",
+    "\n",
+    "where `{start_date}` is a string representing the start of the period we want to read from. \n",
+    "\n",
+    "The stream history for Wikimedia data is not kept indefinitely. Depending on the particular stream configuration, there will likely be between 7 and 31 days of history available. We will choose a past date to start reading data, and stop after 23h of events."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7bc632d9-b47c-4f3d-87fc-3364f90dab4c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from datetime import datetime, timezone, timedelta\n",
+    "\n",
+    "start_time = datetime(2024, 5, 25, 0, 0, 0)\n",
+    "end_time = datetime(2024, 5, 25, 23, 0, 0, tzinfo=timezone.utc)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1ee1bf33-50f8-4256-a003-f38077f888cc",
+   "metadata": {},
+   "source": [
+    "To read this data, we will use a [Pull input adapter](https://github.com/Point72/csp/wiki/Write-Historical-Input-Adapters#pullinputadapter)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9aee0073-6595-47b4-8144-45b6c417af65",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import re\n",
+    "import json\n",
+    "import threading\n",
+    "\n",
+    "from sseclient import SSEClient as EventSource\n",
+    "\n",
+    "import csp\n",
+    "from csp.impl.pulladapter import PullInputAdapter\n",
+    "from csp.impl.wiring import py_pull_adapter_def\n",
+    "\n",
+    "class WikiData(csp.Struct):\n",
+    "    servername: str\n",
+    "    timestamp: str\n",
+    "    event: str\n",
+    "\n",
+    "# Define the runtime implementation of our adapter\n",
+    "class HistoricalWikiDataAdapter(PullInputAdapter):\n",
+    "    def __init__(self, num_ticks: int, url: str):\n",
```

Comment:
its not common practice to ever limit an adapter by number of ticks.
It should run from the requested start time to end time
if you want to end the example after a certain number of ticks it can be done in the graph with
csp.stop_engine() on a condition based on csp.count(event)